app/assets/javascripts/libs/diffable_map.js

```diff
@@ -0,0 +1,294 @@
+// @flow
+// import Utils from "libs/utils";
+
+const defaultItemsPerBatch = 10000;
+let idCounter = 0;
+
+// Some properties:
+// Insertion order is not guaranteed
+
+class DiffableMap<K: number, V> {
+  id: number;
+  chunks: Array<Map<K, V>>;
+  entryCount: number;
+  existsCache: Map<K, boolean>;
+  itemsPerBatch: number;
+
+  constructor(optKeyValueArray?: ?Array<[K, V]>, itemsPerBatch: number = defaultItemsPerBatch) {
+    // Make the id property not enumerable so that it does not interfere with tests
+    Object.defineProperty(this, "id", { value: idCounter++, writable: true });
+    // this.id = idCounter++;
+    this.chunks = [];
+    this.existsCache = new Map();
+    this.entryCount = 0;
+    this.itemsPerBatch = itemsPerBatch;
+
+    if (optKeyValueArray != null) {
+      for (const [key, value] of optKeyValueArray) {
+        this.mutableSet(key, value);
+      }
+    }
+  }
+
+  get(key: K): V {
+    let idx = 0;
+    while (this.chunks[idx] != null) {
+      if (this.chunks[idx].has(key)) {
+        // $FlowFixMe
+        return this.chunks[idx].get(key);
+      }
+      idx++;
+    }
+    throw new Error("Get empty");
+  }
+
+  has(key: K): boolean {
+    try {
+      this.get(key);
+      return true;
+    } catch (exception) {
+      return false;
+    }
+  }
+
+  set(key: K, value: V): DiffableMap<K, V> {
+    if (this.existsCache.has(key)) {
+      let idx = 0;
+      while (this.chunks[idx] != null) {
+        if (this.chunks[idx].has(key)) {
+          const newMap = shallowCopy(this);
+          newMap.chunks[idx] = new Map(this.chunks[idx]);
+          newMap.chunks[idx].set(key, value);
+          return newMap;
+        }
+        idx++;
+      }
+      throw new Error("should not happen");
+    } else {
+      const isTooFull = this.entryCount / this.chunks.length > this.itemsPerBatch;
+      const nonFullMapIdx =
+        isTooFull || this.chunks.length === 0 ? -1 : Math.floor(Math.random() * this.chunks.length);
+
+      // Key didn't exist. Add it.
+      const newDiffableMap = shallowCopy(this);
+      newDiffableMap.existsCache.set(key, true);
+      newDiffableMap.entryCount = this.entryCount + 1;
+      if (nonFullMapIdx > -1) {
+        newDiffableMap.chunks[nonFullMapIdx] = new Map(this.chunks[nonFullMapIdx]);
+        newDiffableMap.chunks[nonFullMapIdx].set(key, value);
+        return newDiffableMap;
+      } else {
+        const freshMap = new Map();
+        freshMap.set(key, value);
+        newDiffableMap.chunks.push(freshMap);
```

Comment:
Couldn't the `DiffableMap` already have a chunk with an empty `Map` when calling the constructor? This could perhaps avoids having to duplicate this code for `set` and `mutableSet`. Also, the "-1" workaround seems odd.

Reply:
The code ensures that the DiffableMap grows dynamically over time. If the "fill degree" is too high, a new chunk is created which is encoded by `-1`. Creating a new set in the constructor would not get rid of this code.
The code duplication could be avoided be extracting it into a separate method, but I was not sure whether the VM would inline this. Should be worth testing, though.